pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go

```diff
@@ -681,90 +833,121 @@ func (p *testPlanner) changeVersionSteps(
 
-	var mixedVersionStepsDuringTenantMigrations []testStep
-	if scheduleHooks {
-		mixedVersionSteps := p.hooks.MixedVersionSteps(p.currentContext, p.prng, p.isLocal)
+	var steps []testStep
 
-		switch p.deploymentMode {
-		case SystemOnlyDeployment:
-			steps = append(steps, mixedVersionSteps...)
+	// runStepsWithTenantMigrations adds a step to `SET` the tenant's
+	// cluster version to the version we are upgrading to. In addition,
+	// if `scheduleHooks` is true, we choose a (possibly empty) subset
+	// of user-hooks to run concurrently with these migrations.
+	runStepsWithTenantMigrations := func() {
+		upgradeSteps := []testStep{
+			p.newSingleStep(setClusterVersionStep{
+				v:                  toVersion,
+				virtualClusterName: p.tenantName(),
+			}),
+		}
 
-		case SharedProcessDeployment:
-			// If we are in a multi-tenant deployment, we run some
-			// mixed-version hooks while the system is finalizing, and the
-			// remaining hooks while the tenant is finalizing.
-			if len(mixedVersionSteps) > 0 {
-				idx := p.prng.Intn(len(mixedVersionSteps))
+		if scheduleHooks {
+			upgradeSteps = append(upgradeSteps, p.hooks.MixedVersionSteps(p.currentContext, p.prng)...)
+		}
 
-				mixedVersionStepsDuringSystemMigrations := mixedVersionSteps[:idx]
-				steps = append(steps, mixedVersionStepsDuringSystemMigrations...)
+		steps = append(steps, p.concurrently(genericLabel, upgradeSteps)...)
+	}
 
-				mixedVersionStepsDuringTenantMigrations = mixedVersionSteps[idx:]
-			}
+	// If the service being upgraded is the system service, we allow
+	// auto-upgrades since they are supported.
+	if service.IsSystem() {
+		steps = []testStep{allowAutoUpgrade}
+		if scheduleHooks {
+			steps = append(
+				steps,
+				p.concurrently(mixedVersionLabel, p.hooks.MixedVersionSteps(p.currentContext, p.prng))...,
+			)
 		}
+	} else {
+		// In separate-process deployments, we set the cluster setting
+		// explicitly, as auto-upgrades are not supported. We also run
+		// all steps selected to run during finalization at this stage.
+		runStepsWithTenantMigrations()
 	}
 
 	steps = append(steps, p.newSingleStep(
 		waitForStableClusterVersionStep{
-			nodes:              p.currentContext.System.Descriptor.Nodes,
+			nodes:              service.Descriptor.Nodes,
 			timeout:            p.options.upgradeTimeout,
 			desiredVersion:     versionToClusterVersion(toVersion),
-			virtualClusterName: install.SystemInterfaceName,
+			virtualClusterName: service.Descriptor.Name,
 		},
 	))
 
-	// At this point, we just upgraded the storage cluster; we might
-	// need to patch up the `system.tenant_settings` table in order for
-	// the tenant upgrade to work.
+	// If this is a system-only or separate-process deployment, we are
+	// done -- the `service` passed is upgraded at this point. However,
+	// for shared-process deployments, we also need to upgrade the
+	// underlying tenant.
+	if p.deploymentMode == SystemOnlyDeployment || p.deploymentMode == SeparateProcessDeployment {
```

Comment:
Nit: I find it confusing to have to switch between `p.deploymentMode`, `p.isMultitenant`, and `virtualClusterSetup`, sometimes in the same func. Perhaps adding to the confusion is the fact that `p.isMultitenant` and `virtualClusterSetup` are _not_ equivalent under all invocations. It's not immediately clear if there is a way to simplify; perhaps a top-level comment summarizing how they're used wrt (FSM) state transitions.

Reply:
Noted.

If it helps:

* `p.deploymentMode` is the actual deployment mode we're generating a plan for. Typically used when we want to do something specific to a single deployment mode (i.e., `SharedProcess` in this case).
* `p.isMultitenant()` is just a helper to group common logic shared by shared- and separate-process multi-tenant deployments.
* `virtualClusterSetup`: admittedly, this is a poorly named variable. It encodes whether the tenant in the test (shared or separate) is "setup" -- i.e., if it's already expected to be running at that point in the plan. Maybe `virtualClusterRunning` is a better name.

Reply:
Renamed `virtualClusterSetup` => `virtualClusterRunning`.